src/main/kotlin/app/RepoHasher.kt

```diff
@@ -90,91 +85,158 @@ class RepoHasher(val localRepo: LocalRepo) {
     }
 
     private fun isKnownRepo(): Boolean {
-        if (Configurator.getRepos().find { it.rehash == repo.rehash } != null) {
-            return true
-        }
-        return false
-    }
-
-    private fun rehashNewCommits() {
-        addedCommits = mutableListOf()
-        removedCommits = repo.commits.toMutableList()
-        val lastServerCommit: Commit = repo.commits.last()
-        var isLastServerCommitChecked: Boolean = false
-
-        val commitsObservable = getObservableRevCommits()
-        commitsObservable.map(commitMapper).filter(emailFilter)
-                .filter { !isLastServerCommitChecked }
-                .blockingSubscribe({  // OnNext
-            Logger.info("Commit: ${it.rehash}")
-            if (it == lastServerCommit) {
-                isLastServerCommitChecked = true
-            }
-            if (removedCommits.contains(it)) {
-                removedCommits.remove(it)
-            } else {
-                addedCommits.add(it)
+        return Configurator.getRepos()
+            .find { it.rehash == repo.rehash } != null
+    }
+
+    private fun findFirstOverlapCommit(): Commit? {
+        val serverHistoryCommits = repo.commits.toHashSet()
+        return getObservableCommits()
+            .skipWhile { commit -> !serverHistoryCommits.contains(commit) }
+            .blockingFirst(null)
+    }
+
+    private fun hashAndSendCommits() {
+        val lastKnownCommit = repo.commits.lastOrNull()
+        val knownCommits = repo.commits.toHashSet()
+        getObservableCommits()
+            .pairWithNext()  // Pair commits to get diff.
+            .takeWhile { (new, _) ->  // Hash until last requested commit.
+                new.rehash != lastKnownCommit?.rehash }
```

Comment:
It's backward. We take after last known commit but we start from latest commit in master and go further. "Requested" is mistake, should be "known".